### ISSUE TITLE
Order-dependent commands

### ISSUE THREAD

**Issue by a-dma** [nix-community/disko#195]
Hi,

I'm trying to put together a slightly generic partition scheme with LUKS+LVM.
The specific layout is two partitions `sda1` for `/boot` and `sda2` for LUKS, with `sda2` containing a single volume group with a fixed amount of storage for swap and "the rest" for root.

Basically, something that looks like this:
```
     sda                     8:0    0    25G  0 disk
     ├─sda1                  8:1    0   511M  0 part  /boot
     └─sda2                  8:2    0  24.5G  0 part
       └─root              254:0    0  24.5G  0 crypt
         ├─vg-swap         254:1    0     4G  0 lvm   [SWAP]
         └─vg-root         254:2    0  20.5G  0 lvm   /
```

I've modified the provided (thanks) LUKS+LVM example to suit my needs, however I run into an issue. The relevant part of the configuration is this:

```
lvm_vg = {
    vg = {
      type = "lvm_vg";
      lvs = {
        swap = {
          type = "lvm_lv";
          size = "4G";
          content = { type = "swap"; };
        };
        root = {
          type = "lvm_lv";
          size = "100%FREE";
          content = {
            type = "filesystem";
            format = "ext4";
            mountpoint = "/";
            mountOptions = [ "defaults" ];
          };
        };
      };
    };
  };
```

which I hoped would translate to the following commands:
```
# lvcreate -L 4G -n swap vg~
# lvcreate -l '100%FREE' -n root vg~
```
allocating 4GB to swap and the remaining space to root. What seems to happen though is that the configuration is evaluated in a different order (I _think_ alphabetical), ultimately flipping around the order of the commands. What that means is that root will be assigned all the available space _first_, and then the swap partition will try to take 4GB of unavailable space. If I rename the swap attrset to "aswap", then that one is processed first and everything works as expected.

I can live with a workaround of renaming things but my question is: is this meant to be a supported behavior? Being able to use `100%FREE` is very convenient. Is there a way to force the evaluation order of the various items without having to resort to specific naming rules?

Thanks!

**Comment by Mic92** [comment]
The issue we encounter here is that nix attrsets are always sorted by name. I think would need to make this a list instead just like this is done for partitions?

**Comment by Lassulus** [comment]
ah, I would just create partitions with 100%FREE size as last. that should fix the issue?

**Comment by a-dma** [comment]
@Mic92 I assume that mapping things to a list would not perform automatic sorting? If that's the case then I think it should work.

@Lassulus do you mean internally treating every `100%FREE` entry as special and "manually" processing those last when resolving them? I think that would also work if it can be done reliably. I can't imagine a case where there's a `100%FREE` entry that is not last, unless of course it's the only entry, in which case it is also the last one. I would imagine this applies to more than just LVS though.

**Comment by Lassulus** [comment]
@a-dma can you check if my PR solves the issue? We could also think about a general solution but special solutions like that should also be useful.

**Comment by a-dma** [comment]
I see the PR has been merged already, so I tested master. Happy to report that it worked for me. Thanks!

The only other thing that I had to do to get my own setup to work was to manually set a label on the swap partition given that there is no `extraArgs` for that type. But that's unrelated and perhaps something that should go into another issue.

Thanks again for the help, closing this.

Resolved via #234.
